draco/asp/hard.lp

```diff
@@ -265,6 +265,84 @@ violation(col_no_x) :-
     entity(mark,V,M),
     not attribute((mark,channel),M,x).
 
+% @hard(stack_without_bar_area) Only use stacking for bar and area.
+violation(stack_without_bar_area) :-
+    attribute(mark_with_stack,M),
+    not attribute((mark,type),M,bar),
+    not attribute((mark,type),M,area).
+
+% @hard(stack_without_summative_agg) Don't stack if aggregation is not summative (summative are count, sum, distinct, valid, missing).
+violation(stack_without_summative_agg) :-
+    entity(encoding,_,E),
+    attribute((encoding,stack),E,_),
+    not attribute((encoding,aggregate),E,sum),
+    not attribute((encoding,aggregate),E,count).
+
+% @hard(no_stack_with_bar_area_discrete_color) Need to stack if we use bar, area with discrete color.
+violation(no_stack_with_bar_area_discrete_color) :-
+    attribute(mark_channel_discrete_or_binned,M,color),
+    attribute((mark,type),M,(bar;area)),
+    not attribute(mark_with_stack,M).
+
+% @hard(stack_without_discrete_color_or_detail) Can only use stack if we also use discrete color, or detail.
+violation(stack_without_discrete_color_or_detail) :-
+    attribute(mark_with_stack,M),
+    not attribute(mark_channel_discrete_or_binned,M,color),
+    not attribute((mark,channel),M,detail).
+
+% @hard(stack_detail_without_q_color) If we use stack and detail, we also have to use quantitative color.
```

Comment:
not sure what `stack_detail_without_q_color` and `stack_detail_without_q_color_agg` are about.
The description says about quantitative, but it's not reflected in the constraints.

Reply:
@domoritz When I translated from Draco1, I was also not sure about these two.

Reply:
I would also have to dig into the meaning. When in doubt, leave it out (and then see when we generate bad visualizations).

Reply:
I removed them for now